File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
        if (!root) return 0;
        int rVal = root->val;
        int left = countGood(root->left, rVal, rVal);
        int right = countGood(root->right, rVal, rVal);
        return  left + right + 1;
    }
private:
    int countGood(TreeNode *node, int rootVal, int mxVal) {
        if (!node) return 0;
        mxVal = max(mxVal, node->val);
        return (node->val >= mxVal) + countGood(node->left, mxVal, mxVal) + countGood(node->right, mxVal, mxVal);
    }
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        queue = deque([])
        curMax = root.val
        queue.append([root, root.val])
        cnt = 1
        while queue:
            sz = len(queue)
            for _ in range(sz):
                r,c = queue.popleft()
                if r.left:
                    if c <= r.left.val:
                        cnt += 1
                    queue.append([r.left, max(c, r.left.val)])
                if r.right:
                    if c <= r.right.val:
                        cnt += 1
                    queue.append([r.right, max(c, r.right.val)])
        return cnt
        
        
        
                